feat(posts): Template versioning article (#450)
BedrockStreaming/tech.bedrockstreaming.com: _posts/2024-08-02-how-to-manage-hundreds-of-templates.md

---
layout: post
title: "How to manage hundreds of view templates?"
description: ""
author: [d_cuny, m_oudji]
category: 
tags: [android, design]
color: rgb(251,87,66)
language: en
comments: true
---

## The context

> Today, at **Bedrock** company, we are providing a streaming service on :

Reviewer: In english, all the ":" are stuck to the previous word, there is no space like in french
```suggestion
Today, at **Bedrock** company, we are providing a streaming service on:
```

- 11 common **CTV** platforms : Tizen, Philips, WebOS, Hisense, Panasonic, Orange, Xbox, PS4, Orsay, Smart Alliance, WhaleOS

> - 5 **Android** platforms : mobile / TV : AndroidTV, bytel, free, SFR, FireTV

Reviewer: Suggestion: I would use another format like for example "mobile / TV (AndroidTV, Bytel, Free, SFR, FireTV)".

How do you count the platforms? What does 5 represent exactly?

- 2 **Apple** platforms : iOS / tvOS
- 1 **Web** platform
- a lot of specific devices

Each of these platforms has its own guidelines and usage (interaction, screen sizes, etc.) and is managed by a dedicated team in our organization. They also have their own native **design components** on which our implementations rely to offer the best possible experience.  

> Bedrock currently have four clients (M6+, RTL+ Hungary, Videoland, Showcase), each of them have their dedicated application on each platform. But they also have their own design expectations and branding guidelines.  

Reviewer: Do we want to mention `Showcase`?

> Bedrock currently have four clients (M6+, RTL+ Hungary, Videoland, Showcase), each of them have their dedicated application on each platform. But they also have their own design expectations and branding guidelines.  

Reviewer: ```suggestion
Bedrock currently has four clients (M6+, RTL+ Hungary, Videoland, Showcase), each of them has their dedicated application on each platform. But they also have their own design expectations and branding guidelines.  
```

> At their disposal, they have a **design system** : a set of visual asset they can use to render pages. These component follow the atomic design model, which means that each layer is a composition of the previous one. Here is a non-exhaustive list : 

Reviewer: ```suggestion
At their disposal, they have a **design system** : a set of visual assets they can use to render pages. These components follow the atomic design model, which means that each layer is a composition of the previous one. Here is a non-exhaustive list : 
```

> - 14 **tokens** : base element that define the brand ex : Colors, FontSystem, Radius, Shadow, Breakpoints, Treatments... 

Reviewer: ```suggestion
- 14 **tokens** : base element that defines the brand (ex: Colors, FontSystem, Radius, Shadow, Breakpoints, Treatments...)
```

> - 9 **atoms** : simplest design item that cannot be decomposed ex : Icons, Avatar, CheckBox, ProgressBar, ServiceIcon, Separator...

Reviewer: ```suggestion
- 9 **atoms** : simplest design item that cannot be decomposed (ex: Icons, Avatar, CheckBox, ProgressBar, ServiceIcon, Separator...)
```

> - 62 **molecules** : group of atoms that form a visual unity ex : Card, Poster, HorizontalCard, Button, CTA, Portrait, Totem...

Reviewer: ```suggestion
- 62 **molecules** : group of atoms that forms a visual unity (ex: Card, Poster, HorizontalCard, Button, CTA, Portrait, Totem...)
```

> - 12 **organisms**: complex visual items or part of the interface : Jumbotron, Hero, Solo, Banner…

Reviewer: ```suggestion
- 12 **organisms**: complex visual item or part of the interface (ex: Jumbotron, Hero, Solo, Banner…)
```

_Sample screens composed of design system components_

:-:|:-:|:-:  
![](/images/posts/2024-08-02-how-to-manage-hundreds-of-templates/home-m6plus.png) | ![](/images/posts/2024-08-02-how-to-manage-hundreds-of-templates/home-videoland.png) | ![](/images/posts/2024-08-02-how-to-manage-hundreds-of-templates/home-rtlhu.png)  

This results in, approximately, one hundred design elements per client. However, not all components are always specific, and an **inheritance** system is in place to allow clients to reuse default atoms, molecules, or organisms while applying the tokens corresponding to their brand.  

The objectives and expectations vary greatly among clients, resulting in numerous graphical element evolutions. Managing the creation and progression of these elements through different processes is a major challenge in tracking designs across each platform. These demands sometimes lead to different integrations. Whether or not the default design is inherited is crucial information to avoid manually comparing the app designs with documentation, which can lead to misunderstandings and wasted time during validation and approval processes.

> ## One versioning to rules them all 

Reviewer: ```suggestion
## One versioning to rule them all 
```

- Stay **up-to-date** with the constant evolutions of the design system to maintain the source of truth.

![Release note of the Design system](/images/posts/2024-08-02-how-to-manage-hundreds-of-templates/design-release-note.png)  
 
Initially, the design team began by versioning its releases and all available components, which are accessible to technical teams in our online design documentation (hosted on [ZeroHeight](https://zeroheight.com/)). This versioning is **common to all platforms** and all clients if there are no specificities for the component in question. Ultimately, versioning helps product and technical teams track the delivery progress of new designs. The version number follows these rules:

Code status | Stage | Rule | Example  
:-:|:-:|:-:|:-:  
First release | New collection | 1.0.0 | 1.0.0  
Documentation update without impact on the anatomy of design element | Documentation update | Increment the third digit | 1.0.1  
New backward-compatible feature | Minor release | Increment the 2nd digit and reset the third | 1.1.0  

> Change on element that break backward compatibility | Major release | Increment the first digit and reset the second and the third | 2.0.0  

Reviewer: ```suggestion
Change on element that breaks backward compatibility | Major release | Increment the first digit and reset the second and the third | 2.0.0  
```

Each frontend team was then responsible for implementing an equivalent versioning system representing the state of their platform. This was done through a **feature team** including all frontends, design system managers, and the product team. This organization brought numerous benefits:

> - Facilitating the **synchronization** and homogeneity of product templates,

Reviewer: ```suggestion
- Facilitating the **synchronization** and homogeneity of product templates
```

> - **Sharing** development challenges,

Reviewer: ```suggestion
- **Sharing** development challenges
```

> - **Limiting** versioning differences between frontends,

Reviewer: ```suggestion
- **Limiting** versioning differences between frontends
```

> - Sharing **industrialization** ideas.

Reviewer: ```suggestion
- Sharing **industrialization** ideas
```

## A concrete sample : Android

> On Android side, we sometimes have different component implementations between mobile and TV, evolving at different paces. This required **two separate versioning** to represent them. Moreover, we currently use comments to track the migration to Google's new view system (Compose), which is happening alongside graphical evolutions.  

Reviewer: ```suggestion
On Android side, we sometimes have different component implementations between mobile and TV, evolving at different paces. This required **two separate versionings** to represent them. Moreover, we currently use comments to track the migration to Google's new view system (Compose), which is happening alongside graphical evolutions.  
```

Code status | Stage | Rule | Example  
:-:|:-:|:-:|:-:  
First release | New collection | 1.0.0 | 1.0.0  
Documentation update without impact on the anatomy of design element | Documentation update | Increment the third digit | 1.0.1  
New backward-compatible feature | Minor release | Increment the 2nd digit and reset the third | 1.1.0  
Change on element that breaks backward compatibility | Major release | Increment the first digit and reset the second and the third | 2.0.0  

Each frontend team were then responsible for implementing an equivalent versioning system representing the state of their platform. This was done through a **feature team** including all front-end teams, design system managers, and the product team. This organization brought numerous benefits:
- Facilitating the **synchronization** and homogeneity of product templates
- **Sharing** development challenges
- **Limiting** versioning differences between all fronts
- Sharing **industrialization** ideas

> ## Case study : Android versioning challenges

Reviewer 2: ```suggestion
## Case study: Android versioning challenges
```

Automatically generated template versioning report for M6+:

![Template versioning report for a customer](/images/posts/2024-08-02-how-to-manage-hundreds-of-templates/template-versioning-report.png)  

## Conclusion

Design system versioning reports offer a **quick and exhaustive overview** of the design system at any given moment for any of our customers and simplifies the QA teams validation work. It is now an essential tool for tracking design progress across all supported platforms.

However, improvements are always possible:
- Currently, developers are responsible for **updating the versioning file** when adding or modifying graphical elements, which can lead to errors and omissions. To avoid this, we would like to link the concrete implementation of the component and its version in the versioning file. But as we are still using the android legacy view system, component can be a class but also some xml style or theme or even resource files. This should be easier after the migration to the new Android view system Compose.

> - The reports are currently generated in **XML format**, which does not facilitate their use; using the PDF format would be preferable.

Reviewer 2: Same comment as I made on TP yesterday, coincidentally; since it's published to GHA now, do we still really need PDFs?
